### DISCUSSION TITLE
Improvements to /server/setup.js

### DISCUSSION THREAD

**Discussion by gbowne1** [gbowne1/codebooker#133] [Ideas]
After doing some research for @Gourav2609 I would like to propose some improvements to /server/setup.js here:

```js
require('dotenv').config({ path: '../.env' });
const bcrypt = require('bcryptjs');
const mongoose = require('mongoose');

mongoose.connect(process.env.MONGO_DB, {
  useNewUrlParser: true,
  useUnifiedTopology: true,
  useCreateIndex: true,
});

mongoose.Promise = global.Promise;

async function createUser() {
  try {
    const User = require('../model/userModel');
    const passwordHash = await bcrypt.hash('test', 12);
    await new User({
      username: 'test',
      email: 'test@gmail.com',
      password: passwordHash,
    }).save({ timeout: 30000 }); // Increase the timeout value if needed
    console.log('👍👍👍👍👍👍👍👍 User created: Done!');
    process.exit();
  } catch (e) {
    console.log('\n👎👎👎👎👎👎👎👎 Error! The Error info is below');
    console.log(e);
    process.exit();
  }
}

createUser();
```

In this code, the timeout property is set to 30,000 milliseconds (30 seconds) for the save() method. You can adjust this value as needed. This will reduce Mongoose timeout errors like what [Gourav2609](https://github.com/Gourav2609) was having in this issue: https://github.com/gbowne1/codebooker/issues/30

@LOGESH-B

**Comment by LOGESH-B** [comment]
Hii @gbowne1 
I think this is better way
![image](https://github.com/gbowne1/codebooker/assets/86541143/de9424f2-e8a7-4e31-b664-87592fb09660)

**Reply by gbowne1** [reply]
the problem is that t [Gourav2609](https://github.com/Gourav2609) was hitting a timer, along with a separate issue, which I read some people have with not using the code like I posted. 

We should also implement 

```js
mongoose.connect(process.env.MONGO_DB, {
  useNewUrlParser: true,
  useUnifiedTopology: true,
  useCreateIndex: true,
});
```

I tested the new setup.js code as above and should eliminate the error @Gorav2609 is getting.  It may be that they are not starting the project in the correct order, especially not starting the mongod/mongo server, etc.  Let's see how they respond.  I would hate to see anyone else have this problem.

**Reply by LOGESH-B** [reply]
I think all seems fine for me...the only thing we need to  wait untill the mongoDb to connect with db..

https://github.com/gbowne1/codebooker/discussions/133#discussioncomment-6776318
So the above thing is enough to overcome this issue🤷‍♂️

**Reply by gbowne1** [reply]
Yes, and it totally depends on the individual users connection.  What might work for both of us may not work for others so we have to think for others as well.  

So by adding a connection process with newUrlParser: true, useunifiedTopology: true, useCreateIndex: true and also adding a timeout is recommended by MongoDB documentation to fix this issue.  We can deal with other issues.  This is why having other people test out the process helps out 100%

We should also document the connection and setup process so people can easily understand in which order to set up and run the project.  What might be simple to us, may not be for someone else.

**Reply by gbowne1** [reply]
I have no problem with the code doing additional error checking or printing to console any additional helpful errors.

**Reply by Gourav2609** [reply]
I strongly agree with @gbowne1  of having a good documentation for the setup when ever a new user tries to collaborate as this as significantly increase contributions for codebooker

the current process is 
![image](https://github.com/gbowne1/codebooker/assets/115522947/3ad419f5-334f-4232-8a2c-1aace9c6ad77)
....it would definitely be better if we could include documentation regarding setup of the project

**Reply by gbowne1** [reply]
We forgot to update this document, the test user has a username of test

**Comment by LOGESH-B** [comment]
Hii @gbowne1 
![image](https://github.com/gbowne1/codebooker/assets/86541143/17ebbcff-ccdb-427c-ab17-abe7487798a7)
I made the mongodb startup service automatic

you can do aswell for better experiance

**Reply by gbowne1** [reply]
yeah thats fine.  Still doesnt solve Gorav's issue. And the issue is not so much as creating a user, it's not giving MongoDB, Mongoose and mongo/mongosh time to interact with each other.  So, changing setup.js to the new setup.js code above would help that out.  Some systems are not gonna be as fast.  I tested this out, it will be fine.. but Gourav might experience some other issue we will have to debug.

**Reply by LOGESH-B** [reply]
https://github.com/gbowne1/codebooker/discussions/133#discussioncomment-6776437

**Comment by gbowne1** [comment]
I think we should also give permissions to the test user as well.  like this:

```js
username: ... ,
email: ... ,
password: ... ,
roles: [
    { role: "readWrite", db: "codebooker" }
]
```

**Reply by gbowne1** [reply]
Ideally that would ensure that the created test user gets full access as just a test user. Any other user would just get standard priveldges.
